src/libopensc/card-piv.c

```diff
@@ -2902,19 +2949,40 @@ piv_finish(sc_card_t *card)
 				free(priv->obj_cache[i].internal_obj_data);
 		}
 		free(priv);
+		card->drv_data = NULL; /* priv */
 	}
 	return 0;
 }
 
-
 static int piv_match_card(sc_card_t *card)
 {
-	int i, k;
+        SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	/* piv_match_card may be called with card->type, set by opensc.conf */
+	/* user provide card type must be one we know */
+	switch (card->type) {
+		case -1:
+		case SC_CARD_TYPE_PIV_II_GENERIC:
+		case SC_CARD_TYPE_PIV_II_HIST:
+		case SC_CARD_TYPE_PIV_II_NEO:
+		case SC_CARD_TYPE_PIV_II_YUBIKEY4:
+			break;
+		default:
+			return 0; /* can not handle the card */
+	}
+	/* its one we know, or we can test for it in piv_init */
+	return 1; /* Let piv_init finish matching */
```

Comment:
Again, you're assuming that `init` will always be followed by `match_card`; don't do it. Please realize that we're discussing this back and forth for weeks now and the SELECT that you want to save here has **nothing** to do with the actual problem you want to solve with this pull request!

Open a seperate pull request or an issue to address this issue and we can discuss how useful `match_card` actually is. But don't delay merging this PR before the next release over this.